### PR TITLE
Refactor lattice code to expose layering and enable easy extension

### PR DESCRIPTION
**Pull request by Keno**
There's been two threads of work involving the compiler's notion of
the inference lattice. One is that the lattice has gotten to complicated
and with too many internal constraints that are not manifest in the
type system. #42596 attempted to address this, but it's quite disruptive
as it changes the lattice types and all the signatures of the lattice
operations, which are used quite extensively throughout the ecosystem
(despite being internal), so that change is quite disruptive (and
something we'd ideally only make the ecosystem do once).

The other thread of work is that people would like to experiment with
a variety of extended lattices outside of base (either to prototype
potential additions to the lattice in base or to do custom abstract
interpretation over the julia code). At the moment, the lattice is
quite closely interwoven with the rest of the abstract interpreter.
In response to this request in #40992, I had proposed a `CustomLattice`
element with callbacks, but this doesn't compose particularly well,
is cumbersome and imposes overhead on some of the hottest parts of
the compiler, so it's a bit of a tough sell to merge into Base.

In this PR, I'd like to propose a refactoring that is relatively
non-invasive to non-Base users, but I think would allow easier
experimentation with changes to the lattice for these two use
cases. In essence, we're splitting the lattice into a ladder of
5 different lattices, each containing the previous lattice as a
sub-lattice. These 5 lattices are:

- JLTypeLattice (Anything that's a `Type`)
- ConstsLattice ( + `Const`, `PartialTypeVar`)
- PartialsLattice ( + `PartialStruct` )
- ConditionalsLattice ( + `Conditional` )
- InferenceLattice ( + `LimitedAccuracy`, `MaybeUndef` )

The idea is that where a lattice element contains another lattice
element (e.g. in `PartialStruct` or `Conditional`), the element
contained may only be from a wider lattice. In this PR, this
is not enforced by the type system. This is quite deliberate, as
I want to retain the types and object layouts of the lattice elements,
but of course a future #42596-like change could add such type
enforcement.

Of particular note is that the `PartialsLattice` and `ConditionalsLattice`
is parameterized and additional layers may be added in the stack.
For example, in #40992, I had proposed a lattice element that refines
`Int` and tracks symbolic expressions. In this setup, this could
be accomplished by adding an appropriate lattice in between the
`ConstsLattice` and the `PartialsLattice` (of course, additional
hooks would be required to make the tfuncs work, but that is
outside the scope of this PR).

I don't think this is a full solution, but I think it'll help us
play with some of these extended lattice options over the next
6-12 months in the packages that want to do this sort of thing.
Presumably once we know what all the potential lattice extensions
look like, we will want to take another look at this (likely
together with whatever solution we come up with for the
AbstractInterpreter composability problem and a rebase of #42596).

WIP because I didn't bother updating and plumbing through the lattice
in all the call sites yet, but that's mostly mechanical, so if we
like this direction, I will make that change and hope to merge this
in short order (because otherwise it'll accumulate massive merge
conflicts).

### CODE REVIEW

**Review by aviatesk — COMMENTED** [review]
I think I like this -- this should allow us to implement base lattice operations with more composability and provide a way to extend them, in a way we keep type stability. 
I guess from external users' point of view, it may still require some understanding of the base lattice design to actually extend it (e.g. we need to understand the implicit order of `InferenceLattice() = ConditionalsLattice(PartialsLattice(ConstsLattice())))` if we want to implement a customized the abstract interpretation without losing the original features), but it'd be much more useful to give an opportunity rather than staying with the current state.

> I will make that change and hope to merge this
in short order (because otherwise it'll accumulate massive merge
conflicts).

Sounds good. I also want to play with this and may add some example test cases on top of it.

File: base/compiler/abstractlattice.jl [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+`ConditionalsLattice{PartialsLattice{ConstsLattice}}`, but wrapped in a struct.
+"""
+struct OptimizerLattice <: AbstractLattice; end
+widen(L::OptimizerLattice) = ConditionalsLattice(PartialsLattice(ConstsLattice()))
+
+"""
+    struct InferenceLattice
+
+The full lattice used for abstract interpration during inference. This takes the
+`OptimizerLattice` and adjoints `LimitedAccuracy` and `MaybeUndef`.
```
Comment (aviatesk) [inline base/compiler/abstractlattice.jl:54]: `MaybeUndef` only appears in optimization, so I think it should be handled by `OptimizerLattice`?

File: base/compiler/abstractlattice.jl [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+struct ConditionalsLattice{L <: AbstractLattice} <: AbstractLattice
+    parent::L
+end
+widen(L::ConditionalsLattice) = L.parent
+
+"""
+    struct OptimizerLattice
+
+The lattice used by the optimizer. Equivalent to
+`ConditionalsLattice{PartialsLattice{ConstsLattice}}`, but wrapped in a struct.
```
Comment (aviatesk) [inline base/compiler/abstractlattice.jl:45]: Technically `Conditional` should never appear during optimization but `InterConditional` can appear as a return type. I'm not sure we want to have additional lattice layer for separating them though.

File: base/compiler/abstractlattice.jl [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+A lattice extending lattice `L` and adjoining `PartialStruct` and `PartialOpaque`.
+"""
+struct PartialsLattice{L <: AbstractLattice} <: AbstractLattice
+    parent::L
+end
+widen(L::PartialsLattice) = L.parent
+
+"""
+    struct PartialsLattice{L}
```
Comment (aviatesk) [inline base/compiler/abstractlattice.jl:32]:
```suggestion
    struct ConditionalsLattice{L}
```

File: base/compiler/abstractlattice.jl (hunk 1 above)
Comment (Keno) [reply base/compiler/abstractlattice.jl:54]: :+1: So I guess `LimitedAccuracy` doesn't show up in the optimizer and `MaybeUndef` doesn't show up in the inference lattice, so I'll make them separate but both wrap the conditionals lattice.

File: base/compiler/abstractlattice.jl (hunk 2 above)
Comment (Keno) [reply base/compiler/abstractlattice.jl:45]: I was thinking about this. Let me work through this some more and see which way looks nicer.

**Comment by Keno** [comment]
@nanosoldier runtests(ALL, vs = ":master")

**Comment by Keno** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by Keno** [comment]
I think this is in reasonable shape. There's a few more things to do, but the bulk of the changes is done, so I'll do the usual PkgEval, etc and then I think we can merge this and do the remaining tweaks as follow-on to minimize the rebase load on other in-progress PRs.

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/46526#issuecomment-1232596830) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1c60614_vs_0a64f5b/report.html).

**Comment by Keno** [comment]
@nanosoldier `runbenchmarks( inference , vs= :master)`

**Comment by Keno** [comment]
@nanosoldier `runbenchmarks("inference", vs= :master)`

**Comment by Keno** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/46526#issuecomment-1232996500) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/1a93697_vs_a83a0e6/report.md).

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/46526#issuecomment-1232993268) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/7b0eeb9_vs_a83a0e6/report.html).

**Comment by Keno** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by Keno** [comment]
@nanosoldier `runtests(["CairoMakie", "CineFiles", "ClusterManagers", "ConceptnetNumberbatch", "CountdownNumbers", "FlameGraphs", "Folds", "Kahuna", "LogicToolkit", "RetroCap", "StrBase", "SuiteSparseGraphBLAS", "TrueRandom"], vs = ":master")`

**Comment by Keno** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/46526#issuecomment-1233209301) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/b4fe1cc_vs_8fa066b/report.html).

File: base/compiler/abstractlattice.jl [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    valid_as_lattice(ti) || return Bottom
+    return ti
+end
+
+"""
+    tmerge(lattice, a, b)
+
+Compute a lattice join of elements `a` and `b` over the lattice `lattice`.
+Note that the computed element need not be the least upper bound of `a` and
+`b`, but rather, we impose some heuristic limits on the complexity of the
```
Comment (vtjnash) [inline base/compiler/abstractlattice.jl:105]: These limits are not heuristic, but rather they are form the strict basis for the definition of `tmerge` here, and are mandatory. In particular, the `tmerge` operation must simultaneously satisfy the lattice join condition over the `lattice` argument, and the global complexity condition, both giving a loose upper bound, which together create a least upper bound under the dual lattice constraints.

Comment (Keno) [reply base/compiler/abstractlattice.jl:105]: Maybe `heuristic` is the wrong word, but I don't think we make any efforts to actually ensure that the result we return is actually the least upper bound under the joint lattice. Computing that is quite hard.

File: base/compiler/abstractlattice.jl [hunk 5]
```diff
@@ -0,0 +1,173 @@
+abstract type AbstractLattice; end
+function widen end
```
Comment (vtjnash) [inline base/compiler/abstractlattice.jl:2]: It feels a little odd to merge this with the existing function of this name
```
help?> widen
  widen(x)

  If x is a type, return a "larger" type, defined so that arithmetic operations + and - are guaranteed not to overflow nor lose precision for any combination of
  values that type x can hold.
```

Comment (Keno) [reply base/compiler/abstractlattice.jl:2]: I can do `widenlattice` instead.

File: base/compiler/abstractlattice.jl [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+const AnyConditionalsLattice{L} = Union{ConditionalsLattice{L}, InterConditionalsLattice{L}}
+const BaseInferenceLattice = typeof(ConditionalsLattice(PartialsLattice(ConstsLattice())))
+const IPOResultLattice = typeof(InterConditionalsLattice(PartialsLattice(ConstsLattice())))
+
+"""
+    struct InferenceLattice{L}
+
+The full lattice used for abstract interpration during inference. Takes
+a base lattice and adjoins `LimitedAccuracy`.
```
[the remark this answers is not in the document]
Comment (Keno) [reply base/compiler/abstractlattice.jl:64]: It had support in the various lattice operations, so I added a layer for it. Of course, it technically violates a whole bunch of lattice assumptions, but that's a different problem. I think it's fine as a lattice layer, though of course the optimizer does not use it.

**Comment by aviatesk** [comment]
As a target example I'm currently writing `TaintInterpreter` that is supposed to do a simple taint analysis using this lattice extension framework. And I found that in addition to several minor fixes on this PR, we will need the following additional efforts for this to work properly:
- pass `AbstractInterpreter` around in the optimization phase so that optimizer can work in the existence of user-extended lattice elements
- make all tfuncs accept `AbstractLattice` allowing external consumers to extend it appropriately

<details><summary> TaintInterpreter </summary>

```julia
using Core: SlotNumber, Argument
using Core.Compiler: slot_id, tmerge_fast_path
import .CC:
    AbstractLattice, BaseInferenceLattice, IPOResultLattice, InferenceLattice, OptimizerLattice,
    widen, is_valid_lattice, typeinf_lattice, ipo_lattice, optimizer_lattice,
    widenconst, tmeet, tmerge, ⊑, abstract_eval_special_value, widenreturn

@newinterp TaintInterpreter
struct TaintLattice{PL<:AbstractLattice} <: CC.AbstractLattice
    parent::PL
end
CC.widen(𝕃::TaintLattice) = 𝕃.parent
CC.is_valid_lattice(𝕃::TaintLattice, @nospecialize(elm)) =
    is_valid_lattice(widen(𝕃), elem) || isa(elm, Taint)

struct InterTaintLattice{PL<:AbstractLattice} <: CC.AbstractLattice
    parent::PL
end
CC.widen(𝕃::InterTaintLattice) = 𝕃.parent
CC.is_valid_lattice(𝕃::InterTaintLattice, @nospecialize(elm)) =
    is_valid_lattice(widen(𝕃), elem) || isa(elm, InterTaint)

const AnyTaintLattice{L} = Union{TaintLattice{L},InterTaintLattice{L}}

CC.typeinf_lattice(::TaintInterpreter) = InferenceLattice(TaintLattice(BaseInferenceLattice.instance))
CC.ipo_lattice(::TaintInterpreter) = InferenceLattice(InterTaintLattice(IPOResultLattice.instance))
CC.optimizer_lattice(::TaintInterpreter) = InterTaintLattice(OptimizerLattice())

struct Taint
    typ
    slots::BitSet
    function Taint(@nospecialize(typ), slots::BitSet)
        if typ isa Taint
            slots = typ.slots ∪ slots
            typ = typ.typ
        end
        return new(typ, slots)
    end
end
Taint(@nospecialize(typ), id::Int) = Taint(typ, push!(BitSet(), id))

struct InterTaint
    typ
    slots::BitSet
    function InterTaint(@nospecialize(typ), slots::BitSet)
        if typ isa InterTaint
            slots = typ.slots ∪ slots
            typ = typ.typ
        end
        return new(typ, slots)
    end
end
InterTaint(@nospecialize(typ), id::Int) = InterTaint(typ, push!(BitSet(), id))

const AnyTaint = Union{Taint, InterTaint}

function CC.tmeet(𝕃::AnyTaintLattice, @nospecialize(v), @nospecialize(t::Type))
    T = isa(𝕃, TaintLattice) ? Taint : InterTaint
    if isa(v, T)
        v = v.typ
    end
    return tmeet(widen(𝕃), v, t)
end
function CC.tmerge(𝕃::AnyTaintLattice, @nospecialize(typea), @nospecialize(typeb))
    r = tmerge_fast_path(𝕃, typea, typeb)
    r !== nothing && return r
    # type-lattice for Taint
    T = isa(𝕃, TaintLattice) ? Taint : InterTaint
    if isa(typea, T)
        if isa(typeb, T)
            return T(
                tmerge(widen(𝕃), typea.typ, typeb),
                typea.slots ∪ typeb.slots)
        else
            typea = typea.typ
        end
    elseif isa(typeb, T)
        typeb = typeb.typ
    end
    return tmerge(widen(𝕃), typea, typeb)
end
function CC.:⊑(𝕃::AnyTaintLattice, @nospecialize(typea), @nospecialize(typeb))
    T = isa(𝕃, TaintLattice) ? Taint : InterTaint
    if isa(typea, T)
        if isa(typeb, T)
            typea.slots ⊆ typeb.slots || return false
            return ⊑(widen(𝕃), typea.typ, typeb.typ)
        end
        typea = typea.typ
    elseif isa(typeb, T)
        return false
    end
    return ⊑(widen(𝕃), typea, typeb)
end
CC.widenconst(taint::AnyTaint) = widenconst(taint.typ)

function CC.abstract_eval_special_value(interp::TaintInterpreter,
    @nospecialize(e), vtypes::CC.VarTable, sv::CC.InferenceState)
    ret = @invoke CC.abstract_eval_special_value(interp::CC.AbstractInterpreter,
        e::Any, vtypes::CC.VarTable, sv::CC.InferenceState)
    if isa(e, SlotNumber) || isa(e, Argument)
        return Taint(ret, slot_id(e))
    end
    return ret
end

function CC.widenreturn(𝕃::InferenceLattice{<:InterTaintLattice}, @nospecialize(rt), @nospecialize(bestguess), nargs::Int, slottypes::Vector{Any}, changes::CC.VarTable)
    if isa(rt, Taint)
        return InterTaint(rt.typ, BitSet((id for id in rt.slots if id ≤ nargs)))
    end
    return CC.widenreturn(widen(𝕃), rt, bestguess, nargs, slottypes, changes)
end

code_typed(ifelse, (Bool, Int, Int); interp=TaintInterpreter())
```

</details>

**Comment by Keno** [comment]
Yes, some of those changes are potential future work, I just wanted to get the lattice infrastructure in place first.

**Comment by Keno** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/46526#issuecomment-1234444787) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/6775206_vs_cb5f401/report.md).
